File: web-ts/src/integers/NInteger.ts
/**
 * NInteger - Integer node type.
 * @JAVA_REF org.modelingvalue.nelumbo.integers.NInteger
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Node } from '../Node';
import type { Functor } from '../patterns/Functor';
import { TokenType } from '../syntax/TokenType';

// Using bigint for arbitrary precision integers (like Java's BigInteger)
const MIN = BigInt(Number.MIN_SAFE_INTEGER);
const MAX = BigInt(Number.MAX_SAFE_INTEGER);

let FUNCTOR: Functor | null = null;

export function setNIntegerFunctor(f: Functor): void {
  FUNCTOR = f;
}

export class NInteger extends Node {
  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    if (args.length === 1 && typeof args[0] === 'string') {
      // Constructor from parsed string
      super(functor, elements, parse(args[0]));
    } else if (args.length === 1 && typeof args[0] === 'bigint') {
      // Constructor from bigint value
      super(functor, elements, args[0]);
    } else {
      super(functor, elements, ...args);
    }
  }

  protected static fromData(data: unknown[], declaration?: Node): NInteger {
    const node = Object.create(NInteger.prototype) as NInteger;
    (node as unknown as { _data: unknown[] })._data = data;
    (node as unknown as { _declaration: Node | undefined })._declaration = declaration ?? node;
    return node;
  }

  protected override struct(data: unknown[], declaration?: Node): Node {
    return NInteger.fromData(data, declaration ?? this.declaration());
  }

  static of(val: bigint): NInteger {
    if (FUNCTOR === null) {
      throw new Error('NInteger FUNCTOR not initialized');
    }
    return new NInteger(FUNCTOR, List(), val);
  }

  value(): bigint {
    return this.get(0) as bigint;
  }

  override toString(_previous?: TokenType[]): string {
    const value = this.value();
    // Use base 36 for very large numbers (like Java's Character.MAX_RADIX)
    const string = value > MAX || value < MIN
      ? `36#${value.toString(36)}`
      : value.toString();

    if (_previous && (_previous[0] === TokenType.NAME || _previous[0] === TokenType.NUMBER || _previous[0] === TokenType.DECIMAL)) {
      _previous[0] = TokenType.NUMBER;
      return ' ' + string;
    }
    if (_previous) {
      _previous[0] = TokenType.NUMBER;
    }
    return string;
  }
}

function parse(string: string): bigint {
  const i = string.indexOf('#');
  if (i > 0) {
    const radix = parseInt(string.substring(0, i), 10);
    return BigInt(parseInt(string.substring(i + 1), radix));
  }
  return BigInt(string);
}

File: src/main/java/org/modelingvalue/nelumbo/integers/NInteger.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.integers;

import java.io.Serial;
import java.math.BigInteger;

import org.modelingvalue.nelumbo.ConstructionReason;
import org.modelingvalue.nelumbo.KnowledgeBase;
import org.modelingvalue.nelumbo.NelumboConstructor;
import org.modelingvalue.nelumbo.NelumboFunctorField;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.NodeInfo;
import org.modelingvalue.nelumbo.lang.Functor;
import org.modelingvalue.nelumbo.syntax.ParseContext;
import org.modelingvalue.nelumbo.syntax.ParseException;
import org.modelingvalue.nelumbo.syntax.TokenType;

public final class NInteger extends Node {
    @Serial
    private static final long serialVersionUID = 2454372545442550574L;

    private static final BigInteger MIN = BigInteger.valueOf(Long.MIN_VALUE);
    private static final BigInteger MAX = BigInteger.valueOf(Long.MAX_VALUE);

    @NelumboFunctorField
    private static Functor FUNCTOR;

    public static NInteger of(BigInteger val) {
        return new NInteger(NodeInfo.of(FUNCTOR), val);
    }

    @NelumboConstructor
    public NInteger(NodeInfo nodeInfo, Object... args) {
        super(nodeInfo, args);
    }

    private static BigInteger parse(String string) {
        int i = string.indexOf('#');
        if (i > 0) {
            int radix = Integer.parseInt(string.substring(0, i));
            return new BigInteger(string.substring(i + 1), radix);
        }
        return new BigInteger(string);
    }

    @Override
    public String toString(TokenType[] previous) {
        BigInteger value = getVal(0);
        String string = value.compareTo(MAX) > 0 || value.compareTo(MIN) < 0
                ? (Character.MAX_RADIX + "#" + value.toString(Character.MAX_RADIX))
                : value.toString();
        if (previous[0] == TokenType.NAME || previous[0] == TokenType.NUMBER) {
            previous[0] = TokenType.NUMBER;
            return " " + string;
        }
        previous[0] = TokenType.NUMBER;
        return string;
    }

    @Override
    public Node init(KnowledgeBase knowledgeBase, ParseContext ctx, ConstructionReason reason) throws ParseException {
        if (reason == ConstructionReason.parsing && length() == 2 && get(1) instanceof String string) {
            if ("-".equals(get(0))) {
                string = "-" + string;
            }
            BigInteger val = parse(string);
            return setArgs(new Object[] { val });
        }
        return this;
    }

}
